Bug 1471316-support getting credentials from tc-pulse

Biboswan opened the pull request:
Oh didn't release I was working on main repo. @djmitche have a look. There is some code repeat. Maybe you would like to refactor it.

Biboswan committed:
Include taskcluster client

Biboswan committed:
Initial implementation:Bug 1471316 - support credentials from tc-pulse

djmitche commented on src/client.js:
```diff
@@ -105,6 +123,25 @@ class Client extends events.EventEmitter {
+          this.connectionString = pulsecredentials.connectionString;
+          this.namespace = pulsecredentials.namespace;
+          this.reclaimAt = pulsecredentials.reclaimAt;
+          this.recycle();
```
Here, too, it'd be better to just call `recycle` in the interval, knowing that when the instance eventually tries to reconnect, `reclaimAt` will be in the past and it will fetch a new connection string.

djmitche commented on src/client.js:
```diff
@@ -67,19 +68,36 @@ exports.buildConnectionString = buildConnectionString;
+      });
+      this.namespace = username;
+      this.contact = contact;
+      const pulsecredentials = pulseCredentials();
```
This is an async function, so you need to call it with `await`.  But it's too early to call it here -- rather, it should be called just before connecting, and only if necessary.

djmitche commented on src/client.js:
```diff
@@ -105,6 +123,25 @@ class Client extends events.EventEmitter {
+            () => this.recycle(),
+            this._recycleInterval);
+          setTimeout(callreclaim, this.reclaimAt-taskcluster.fromNow('0 minutes'));
+        },
```
I like this :)

djmitche commented on src/client.js:
```diff
@@ -166,6 +203,16 @@ class Client extends events.EventEmitter {
+   */
+  async pulseCredentials() {
+    return await this.pulse.claimNamespace(this.namespace, {
+      expires: taskcluster.fromNow('1 day'),
```
This should probably be configurable.

djmitche reviewed (changes requested):
Nice start!

This is starting to become a lot of options to the constructor:
 * username/password/host/vhost; OR
 * connectionString; OR
 * credentials, rootUrl, contact, expires

In all of these cases, all the Client needs is (a) a way to get a connectionString when it needs one (when connecting) and (b) the time until that connectionString is invalid and it needs to recycle.

So, it might make sense to make three new small functions or classes, one for each of the above cases, and require that one of those three be passed to the constructor.

```javascript
new Client({..., credentials: pulseCredentials({username, password, hostname, vhost})});
// or
new Client({..., credentials: connectionStringCredentials('...')});
// or
new Client({..., credentials: claimedCredentials({rootUrl, credentials, expires, contact})})
```

Each of those functions would return an async function that returns `{connectionString, recycleAfter}`.  For the first two, the function would be pretty simple (just return a static connection string and no `recycleAfter`) and for the third option, it would call the taskcluster-pulse service.

In a more object-oriented language, we might make three different classes in a hierarchy with a `getCredentials` method returning the same thing, but that isn't really the JS way to do things.

P.S. What happened with all the commits in this PR??

Biboswan commented:
the constructor parameters look messy better to separate the credential handling part, it was in my mind.

Biboswan commented:
@djmitche  You would rather prefer it like `` new Client( ...,  pulseCredentials,connectionStringCredentials,  claimedCredentials) ``. I mean not to pass functions in the constructor just look for the credentials type and  act accordingly.

djmitche commented:
I don't prefer that solution because even in that case, Client must have a case for each credential-getting method.  If it just takes a single argument, then it only has one case.  And later we can add another credential-getting method without modifying Client.

Biboswan committed:
Include different functions to fetch connectionstring

Biboswan committed:
Include credentials.js for export

djmitche commented on src/credentials.js:
```diff
@@ -0,0 +1,68 @@
+      contact,
+    });
+    const connectionString = res.connectionString;
+    const recycleAfter = res.reclaimAt-taskcluster.fromNow('0 minutes');
```
space around `-`

djmitche commented on src/index.js:
```diff
@@ -1,7 +1,9 @@
 module.exports = {
   Client,
   consume,
+  credentials,
```
As I understand it, it would look like

```javascript
const {Client, credentials} = require('taskcluster-lib-pulse');

...

new Client({
  credentials: credentials.claimedCredentials({ ..}),
  ...
});
```

that "stutters" a little -- the word "credentials" appears several times.  I think it would be better to export the individual credential functions at the top level of the module, so you could write

```javascript
const {Client, claimedCredentials} = require('taskcluster-lib-pulse');
```

Biboswan committed:
Add whitespace around '-'

Biboswan committed:
Export each function in credentials.js  as a module

Biboswan committed:
Export each function in credentials.js  as a module

Biboswan committed:
Refactor code with credentials implementation

Biboswan committed:
Fix export

Biboswan committed:
Add catch block for setConnStringNamespace

djmitche reviewed (commented):
This is looking close!

The new code in the Connection.connect method is, I think, something that should be handled in the Client instead.  One of the hints to this is that it's setting `this.client._foo` internal variables.  Instead, you might think about creating a `Client.getConnectionString` async method which, if it has a current connectionString, just returns it, otherwise calls the credential function and uses the result (and updates its internal variables).

Regarding `Client.namespace`, I don't know why I didn't just make that a mandatory Client constructor argument, and resorted to all of this trickery of parsing namespaces out of URLs.  Please remove that and just require that the user pass a `namespace` option to the `Client` constructor.

Biboswan committed:
Fix typo

Biboswan committed:
Implement new credentials function in client test

Biboswan commented:
Modifying client tests. All tests are not passing

Biboswan commented:
oh ! I didn't see your previous comment

Biboswan commented:
namespace is equivalent to username always?

Biboswan committed:
Make namespace parameter for client mandatory

Biboswan committed:
Include getConnectionString function property in client

Biboswan committed:
Correct namespace

Biboswan committed:
Implement connectionStringCredentials for connectionstring

Biboswan commented:
@djmitche  time for new tests I think :)

djmitche commented:
For connection strings returned from tc-pulse, no, they are not the same -- tc-pulse invents several usernames for the same namespace and alternates between them.

Biboswan committed:
Resolve merge conflict

Biboswan committed:
Refactor code

Biboswan committed:
Add mock for tc-pulse and test for reclaim

Biboswan committed:
Include reclaim test in connectionTests

Biboswan commented:
have a look

djmitche commented:
I've rebased this PR onto master.

djmitche commented on src/credentials.js:
```diff
@@ -0,0 +1,82 @@
+ * {
+ *   username:          // Pulse username
+ *   password:          // Pulse password
+ *   hostname:          // Hostname to use
```
and vhost :)

djmitche commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "taskcluster-lib-pulse",
-  "version": "1.1.0",
+  "version": "2.0.0",
```
Please revert this bit.  I will do a 2.0.0 release on master after merging, at which time I'll update package.json, make a git tag, and upload the release.

imbstack replied:
Agreed, please undo this change.

djmitche commented on src/credentials.js:
```diff
@@ -0,0 +1,82 @@
+};
+
+exports.pulseCredentials = pulseCredentials;
+
```
A short docstring here would be good, for symmetry with the other options.

djmitche commented on src/credentials.js:
```diff
@@ -0,0 +1,82 @@
+  };
+};
+
+exports.mockclaimedCredentials = mockclaimedCredentials;
```
capitalization: `mockClaimed..`

djmitche commented on test/client_test.js:
```diff
@@ -15,36 +14,40 @@ if (!PULSE_CONNECTION_STRING) {
 }
 
 suite('buildConnectionString', function() {
```
this is now testing `pulseCredentials`, so it should probably be in `test_credentials.js`.

djmitche commented on test/client_test.js:
```diff
@@ -319,21 +333,38 @@ const connectionTests = connectionString => {
-      assume(() => new Client({monitor}))
-        .throws(/is required/);
-    });
+  test('reclaim', async function() {
```
This is testing `Client` functionality, so it belongs in the `Client` suite, below.

djmitche commented on src/client.js:
```diff
@@ -237,6 +189,25 @@ class Client extends events.EventEmitter {
+      await this.setConnectionString().catch(err => {});
+      this._timeoutId = setTimeout(reclaim, this._recycleAfter);
+    }, this._recycleAfter);
+  }
```
This looks a bit complex!  I think it will not set `this._timeoutId` until the timeout has expired once, for example.  It also assumes that `recycleAfter` is a repeating interval, but it's not -- it's just a one-time thing.

I think there are two pieces to this recycling which, together, will have the desired effect:
 * Make sure we call `client.recycle()` at `credentials.recycleAfter` ms after the last `this.credentials` call.
 * Make sure that we call `this.credentials` if we need a connection string after the `recycleAfter` period has elapsed.
The total effect is this: the Client will recycle at the appropriate time, causing a new connection.  When that new connection goes looking for credentials, the credentials will be out of date, so it will call `this.credentials()` again.

Here's how I'd organize that:
 * have credentials return `recycleAt` (a Date) instead of `recycleAfter`
 * make a `Client.getConnectionString` method which will either return a cached `connectionString` or, if there is no connection string or the time is after `this.recycleAfter`, call `this.credentials`, cache the result, reset the timer, and return it.
 * Make a `Client._setRecycleAt(recycleAt)` that manages `this._recycleTimer`: if any timer exists, clear it, then use `setTimer` to make a new one that will call `this.recycle()` at `recycleAt` (but not loop).

Biboswan replied:
I used the `this._timeoutId` to kill the latest timout with`clearTimeout` along with clearing the `recycle` interval when I stop the  `client`. 
 Further you wanted to make client as much independent of method of fetching credentials as possible so I assume that `recycleAfter` may change next time we call pulse service or any other. So it may not be a one time thing.
Previously I think you told to remove the recycleinterval from reclaim function so i did it :)

djmitche replied:
Agreed, `recycleAfter` will be different every time, so an interval (repeating callback) isn't the right choice, but what you've implemented is a similar thing to setInterval, just implemented with setTimeout (that is, the callback for the first timeout calls setTimeout to set the next one..).

Regarding recycleInterval, yes, I think we can get this working without it, then maybe in a follow-up add it back in.

Biboswan replied:
If you notice https://github.com/taskcluster/taskcluster-lib-pulse/blob/get-pulse-credential/src/client.js#L286 if there is no `recycleAfter`,  `setConnectionString` will be called only first time (same effect as caching). 
Also `setTimeout` returns id at initiation.
If I try to organize code as you said, then I need to use `recycleAt(date)` to judge whether setConnection was called due to `recycleTimer` and not `recycleInterval` and also use Taskcluster client library to calculate correct time interval since new Date() might be system dependent. 
Once setTimeout is called it will clear by itself once finished. Why to clear it anyway.
I can get connectionString anytime with `this.client.connectionString`
I feel present logic is less complicated :) Don't know if I missed anything

djmitche replied:
I think the correctness issue here is that `callreclaim` sets up a loop of calling `setConnectionString`, but does not actually recycle the client when the connection string changes.  So an existing connection would continue using the old connectionString even after a new connectionString had been fetched.  Also, the first return value from setTimeout in `callreclaim` isn't assigned to anything -- so if that timeout hasn't expired, calling `client.stop` will not actually stop the client.

That said, code can be "correct" (does the necessary things) and still need changes for organizational reasons.  Here I see that the correctness of the overall implementation depends on some careful coordination between the Client and Connection classes.  Connection expects setConnectionString to set client.connectionString as well as (sometimes) client._recycleAfter (an internal attribute).  Then Client relies on Connection to call client.callreclaim, and that method depends on setConnectionString to have set this._recycleAfter in that case (which it probably would have since callreclaim doesn't get called unless the first call to this.credentials returned a recycleAfter, and as we have things implemented now that means the second call to this.credentials will, too -- but that behavior isn't part of the contract for this.credentials!)

Looking at the bigger picture of the coordination of Client and Connection:
 * only the Client ever creates a Connection
 * a Connection only uses its connectionString at connection time -- it never reconnects
 * the Client retires Connections when they should no longer be used

So, Connection shouldn't know anything about this claiming stuff.  It already takes a connectionString and connects to it, and should not need to be changed.  Well, almost -- https://github.com/taskcluster/taskcluster-lib-pulse/blob/get-pulse-credential/src/client.js#L286 is a bit of a misdesign, as really the Client should just be passing a connectionString in to Connection.connect(connectionString), instead of the Connection looking at `this.client.connectionString`.  That was my fault :)

Client needs to do two things:
 * ask for a connectionString just before connecting
 * stop using connections at `recycleAt`

For the first, we can just call `await this.credentials()` right before `newConn.connect(connectionString)`.  We could add some caching to that value (with the result valid until `recycleAt`), using a private Client method like `_fetchCredentials`.  Or, that caching could be implemented in `claimedCredentials`.

For the second, we need to call `this.recycle()` at `recycleAt` -- that's it!  The recycling process will then take care of retiring any connections, setting up a new connection, calling `this.credentials()`, and starting a timer to call `this.recycle()` again.  Managing that timer could be handled in the same `_fetchCredentials` method.

I think a design like this will be more maintainable and grow with the system as we add more functionality.  It's less surprising or confusing to read, which helps the next person successfully change things without introducing bugs.

djmitche reviewed (changes requested):
This is looking good!  Since I rebased, you'll want to `git pull --rebase` before you start hacking again.

The two big themes are:
 * fixing up the recycling timer
 * `test_credentials.js` (really just moving some tests into that file)

djmitche commented:
I'm going on vacation for the next two weeks.  Hopefully @imbstack can have a look at the next version?

Biboswan commented:
@imbstack  I think `recycle` method needs to be called seperately with a `Client.initialze()`, because as per previous discussion with @djmitche  we call `fetchCredentials` inside recycle making it `async` which may fail (reject) if pulse service rejects. Anything better ?

imbstack commented:
@Biboswan can you rebase onto master? I'll need to read more to see about `recycle`.

Biboswan committed:
Merge branch 'master' into get-pulse-credential

Biboswan committed:
Fix merge-conflict error

Biboswan committed:
Incorporate new client in publisher and consumer test

Biboswan commented:
Resolved conflict  @imbstack

imbstack commented on src/client.js:
```diff
@@ -3,39 +3,8 @@ const debug = require('debug');
 const amqplib = require('amqplib');
 const assert = require('assert');
 const {URL} = require('url');
-
```
You'll want to leave this extra line here!

imbstack commented on src/client.js:
```diff
@@ -3,39 +3,8 @@ const debug = require('debug');
 const assert = require('assert');
 const {URL} = require('url');
-
 var clientCounter = 0;
```
May as well make this a `let` while we're at it.

Biboswan replied:
hm I have thought on that :)

imbstack commented on src/client.js:
```diff
@@ -66,38 +31,24 @@ exports.buildConnectionString = buildConnectionString;
 class Client extends events.EventEmitter {
-  constructor({username, password, hostname, vhost, connectionString, recycleInterval,
-    retirementDelay, minReconnectionInterval, monitor}) {
+  constructor({namespace, recycleInterval, retirementDelay, minReconnectionInterval, monitor, credentials}) {
```
Since we're changing the interface to this package (this class is exported externally, right?) can we still take `connectionString` as an argument for now and inside the constructor have:

```js
assert(!connectionString, 'connectionString is deprecated');
```

Biboswan replied:
Good idea. But even `namespace`, `password`,  `hostname`, and`vhost`.  are deprecated. Infact  user is more likely to use these parameters for building connection I think. Therefore shall we do that for each such deprecated parameter ?

imbstack replied:
Yep! Sounds like a great plan.

imbstack commented on src/client.js:
```diff
@@ -66,38 +31,24 @@ exports.buildConnectionString = buildConnectionString;
     this.id = ++clientCounter;
     this.debug = debug(`taskcluster-lib-pulse.client-${this.id}`);
-
+  
```
Extra whitespace

imbstack commented on src/client.js:
```diff
@@ -111,7 +62,8 @@ class Client extends events.EventEmitter {
     this.debug('stopping');
     this.running = false;
-
+  
```
I think this might be extra whitespace as well? You can probably set up whatever editor you're using to show whitespace at the end of the line.

imbstack reviewed (commented):
This is looking great so far! I agree with @djmitche's thoughts on `recycle` and think that if you need an `initialize` that is fine!

Biboswan committed:
seperate credentials test in a new file

Biboswan committed:
claimedCredentials caches credentials with a valid expiry date

Biboswan committed:
client in the api is now an async function
Logic is similar to consumer. client creates a new Client instance and then calls start to setup connection and recycleinterval
Incorporate changes to existing tests

Biboswan commented:
@imbstack  updated! there has been significant refactoring. There is just one `recycle test` in client for which I would like to have a better workaround.

Biboswan committed:
Change version to current one

imbstack commented on src/client.js:
```diff
@@ -66,41 +31,42 @@ exports.buildConnectionString = buildConnectionString;
-    } else {
-      this.connectionString = buildConnectionString({username, password, hostname, vhost});
-      this.namespace = username;
+    if (options) {
```
I don't think people passed in an `options` object before. So this would be  just 

```js
assert(!username, 'username is deprecated');
...
```
instead. Does that make sense to you?

Biboswan replied:
yeah, i messed it.

imbstack commented on src/client.js:
```diff
@@ -237,8 +224,30 @@ class Client extends events.EventEmitter {
+  }
 }
 
+const client = async (options) => {
```
I think I'd prefer not to have this and just force people to call `start()` themselves directly. If you really want to keep this, it should be called something like `createClient` or something along those lines.

Biboswan replied:
I think we can retain this with a better name like `createClient` for some symmetry, since in consumer `start` is called internally too like this.

imbstack commented on src/client.js:
```diff
@@ -66,41 +31,42 @@ exports.buildConnectionString = buildConnectionString;
 
+    assert(namespace, 'namespace is required');
+    this.namespace = namespace;
+    this.credentials = credentials;
```
afaict, `credentials` is required as well, right? Let's assert it.

imbstack commented on src/client.js:
```diff
@@ -66,41 +31,42 @@ exports.buildConnectionString = buildConnectionString;
+  * 
+  * In the public API, this is called automatically by `client`
+  */
+  async _start() {
```
Let's call this `start`

imbstack commented on src/client.js:
```diff
@@ -137,23 +104,43 @@ class Client extends events.EventEmitter {
-      });
-      this.connections.unshift(newConn);
+      // If recycle is called due to recycleAt then refresh the normal recycle interval
+      if (this.recycleAt && this.recycleAt - new Date() === 0) {
```
This will almost never be true. Time will pass between when `recycle` is called and you construct your new `Date`. I think you actually shouldn't make this check anyway. The old way with `setTimeout` seems to make more sense to me.

imbstack replied:
There's still more work to do in this function to line up with what @djmitche was saying in [this comment](https://github.com/taskcluster/taskcluster-lib-pulse/pull/7#discussion_r205481788). For instance `Connection` should not take a retirement delay at all, right? That is entirely managed by the client.

Biboswan replied:
Ah that's my fault. @djmitche  wanted to drop the callback effect with `setTimeout` and seperately handle that logic. I think something like `this.recycleAt - new Date() > 5 && this.recycleAt - new Date() < 15` can make sense. Do same in  `this.claimedCredentials` too.

Biboswan replied:
@djmitche  I wanted to know how's the above mentioned logic

djmitche replied:
I agree -- comparing a time for equality will never be true.

Biboswan replied:
@djmitche equality will never be true but  can we do something like `this.recycleAt- new Date` is in interval [-15,15]ms

djmitche replied:
That's still assuming things happen within 15ms, and when they don't it might cause a failure.  It's best to avoid this kind of situation entirely.  I think my comments in the review are a way to do so?

imbstack reviewed (commented):
I think we're 90% of the way there. Just need to get `recycle` up to speed!

djmitche commented:
@Biboswan I'm back from vacation... how is this going?

djmitche commented on src/client.js:
```diff
@@ -137,23 +104,43 @@ class Client extends events.EventEmitter {
+        this.connections.unshift(newConn);
+      } catch (err) {
+        return err;
+      }
```
Returning an error (instead of throwing it) is pretty unusual.  I guess this is to catch errors from fetchCredentials?  Let's handle that directly..

djmitche reviewed (commented):
Let's simplify the concern with timers as follows: treat `_recycleInterval` as providing a default value for `recycleAfter`, if `credentials` doesn't provide a value.  Then there's no need to behave differently if recycle was called due to recycleAt.  So we can remove `this._interval` entirely (since we won't be using setInterval anymore) and just focus on `this._recycleTimer`, which is cleared and reset to `recycleAt` (possibly with a default value) on every credential fetch.

I see that you've added a `client` function and an async `start` method and made `recycle` async, I think because it must now await `this._fetchCredentials()`.  This opens up a bit of a race condition, though: what if `recycle` is called twice in quick succession, so that the second call occurs before `this._fetchCredentials()` has completed for the first call?

I think we can go back to how it was: `new Client(..)` is enough to create a new client which will start connecting immediately; and `client.recycle()` is sync (and will result in a new connection later, after it returns).  The trick is, we really don't need the connectionString until `newConn.connect()` is called -- and we already wait and call that later.  So, something like this:

```javascript
const newConn = new Connection(this._retirementDelay); 

// don't actually start connecting until at least minReconnectionInterval has passed
// and we have a connection string
const earliestConnectionTime = this.lastConnectionTime + this._minReconnectionInterval;
const now = new Date().getTime();
setTimeout(async () => {
  try {
    this.lastConnectionTime = new Date().getTime();
    const connectionString = await this._fetchCredentials();
    newConn.connect(connectionString);
  } catch (err) {
    this.debug(`Error while fetching credentials: ${err}`);
    this.recycle();
  }
}, now < earliestConnectionTime ? earliestConnectionTime - now : 0);    
```

and change the `_fetchCredentials` method to handle `recycleAt` and `this._recycleTimer` and applying a default value based on `recycleInterval`.

Biboswan commented:
@djmitche  I 'm not sure why to give `recycleAfter`  a default value. If the service returns a `recycleAt` then it would do so everytime or else` setInterval` will be there to call `recycle`
As per your review I get a logic like this inside  `_fetchCredentials` 
```javascript
const {connectionString, recycleAt} = await this.credentials();
    if (recycleAt) {
      const recycleAfter = recycleAt - new Date();
      this._recycleTimer = setTimeout(() => this.recycle(), recycleAfter);
      if (this._interval) {
        clearInterval(this._interval);
        this._interval = null;
      }
    }
    
    if (!recycleAt && !this._interval) {
      this._recycleTimer = setTimeout(() => this.recycle(), this._recycleInterval);
    }
 ```
So do you want this to be in a super safe side? then ok I will do this.

djmitche commented:
I think we should remove the use of setInterval entirely.  So if a service returns recycleAt, then `this._recycleInterval` is totally ignored.

Biboswan committed:
Made recycle sync again
and keep one time of timer for recycle

Biboswan commented:
check!

djmitche reviewed (approved):
This looks good!

djmitche commented:
Brian, if you're happy, let's merge :)

Biboswan commented:
First lets update the readme :)
